Yarn 2 PnP ERR_MODULE_NOT_FOUND in SvelteKit

### Describe the bug

The extension throws the following error with yarn 2 PnP and SvelteKit:
```
Error [ERR_MODULE_NOT_FOUND]: Cannot find package '@sveltejs/adapter-auto' imported from /home/haowenl/github/lxy-portfolio/svelte.config.js
    at new NodeError (node:internal/errors:372:5)
    at packageResolve (node:internal/modules/esm/resolve:954:9)
    at moduleResolve (node:internal/modules/esm/resolve:1003:20)
    at defaultResolve (node:internal/modules/esm/resolve:1218:11)
    at ESMLoader.resolve (node:internal/modules/esm/loader:580:30)
    at ESMLoader.getModuleJob (node:internal/modules/esm/loader:294:18)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:80:40)
    at link (node:internal/modules/esm/module_job:78:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

### Reproduction

In a brand new skeleton svelte project, follow these steps: https://github.com/sveltejs/language-tools/tree/master/packages/svelte-vscode#usage-with-yarn-2-pnp.

### Expected behaviour

The extension works as normal.

### System Info

- OS: Fedora 36, Node v16
- IDE: VSCode


### Which package is the issue about?

Svelte for VS Code

### Additional Information, eg. Screenshots

I'm not sure this is correct, but I think the node backend shouldn't be using the internal module loader, as shown in the error message above.
To resolve it, I think it should be sufficient to be able to add `-r .pnp.cjs --loader .pnp.loader.mjs` generated by yarn, as shown in https://github.com/yarnpkg/berry/issues/638#issuecomment-1124608560.

https://github.com/sveltejs/language-tools/blob/5a6eba15918a556e2bcb04638c313ac9dd03212c/packages/svelte-vscode/src/extension.ts#L102

Hopefully, an option alongside `svelte.language-server.runtime`, like `svelte.language-server.runtimeArgs` should be sufficient for this.

Other cases of people encountering this: https://github.com/sveltejs/language-tools/issues/678#issuecomment-1046351495.

Alright, I got a `runtimeArgs` implementation working.

The main issue is that unfortunately it seems to require using an absolute path because it can't find `.pnp.cjs` if I just add the flags `-r .pnp.cjs --loader .pnp.loader.mjs`.

I tried setting the `serverOptions.run.options.cwd` and `serverOptions.debug,options.cwd` both to `rootPath` which indeed is the root of my project so `[rootPath]/.pnp.cjs`. It just fails with `Error: Cannot find module '.pnp.cjs'` unless I use an absolute path.

I don't think that's tenable to add to the `.vscode/settings.json` because per project it needs to be relative and per user they might work on different projects with different paths.

Here's the commit that I made to my own repository to add this change:
https://github.com/DavidArchibald/language-tools/commit/723966447c2d6ae9ca4d032532151f41c127bcc9

Any advice on how to work around this problem?

--- a/packages/svelte-vscode/README.md
+++ b/packages/svelte-vscode/README.md
@@ -64,6 +64,11 @@ This setting can only be changed in user settings for security reasons.
 You normally don't set this. Path to the language server executable. If you installed the `svelte-language-server` npm package, it's within there at `bin/server.js`. Path can be either relative to your workspace root or absolute. Set this only if you want to use a custom version of the language server. This will then also use the workspace version of TypeScript.
 This setting can only be changed in user settings for security reasons.
 
+#### `svelte.language-server.runtime-args`
+
+You normally don't set this. Additional arguments to pass to Node when spawning the language server.
+This is useful when you use something like Yarn PnP and need its loader arguments `["--loader", ".pnp.loader.mjs"]`.
+
 ##### `svelte.language-server.port`
 
 You normally don't set this. At which port to spawn the language server.
@@ -86,7 +91,7 @@ Settings to toggle specific features of the extension. The full list of all sett
 ### Usage with Yarn 2 PnP
 
 1. Run `yarn add -D svelte-language-server` to install svelte-language-server as a dev dependency
-2. Run `yarn dlx @yarnpkg/sdks vscode` to generate or update the VSCode/Yarn integration SDKs. This also sets the `svelte.language-server.ls-path` setting for the workspace, pointing it to the workspace-installed language server. Note that this requires workspace trust - else set the `svelte.language-server.ls-path` setting in your user configuration.
+2. Run `yarn dlx @yarnpkg/sdks vscode` to generate or update the VSCode/Yarn integration SDKs. This also sets the `svelte.language-server.ls-path` and `svelte.language-server.runtime-args` setting for the workspace, pointing it to the workspace-installed language server. Note that this requires workspace trust - else set the `svelte.language-server.ls-path` and `svelte.language-server.runtime-args` setting in your user configuration.
 3. Restart VSCode.
 4. Commit the changes to `.yarn/sdks`
 

--- a/packages/svelte-vscode/package.json
+++ b/packages/svelte-vscode/package.json
@@ -50,7 +50,8 @@
             "supported": "limited",
             "restrictedConfigurations": [
                 "svelte.language-server.runtime",
-                "svelte.language-server.ls-path"
+                "svelte.language-server.ls-path",
+                "svelte.language-server.runtime-args"
             ],
             "description": "The extension requires workspace trust because it executes code specified by the workspace. Loading the user's node_modules and loading svelte config files is disabled when untrusted"
         }
@@ -87,6 +88,11 @@
                     "type": "string",
                     "title": "Language Server Path",
                     "description": "- You normally don't set this - Path to the language server executable. If you installed the \"svelte-language-server\" npm package, it's within there at \"bin/server.js\". Path can be either relative to your workspace root or absolute. Set this only if you want to use a custom version of the language server. This will then also use the workspace version of TypeScript. This setting can only be changed in user settings for security reasons."
+                },
+                "svelte.language-server.runtime-args": {
+                    "type": "array",
+                    "title": "Language Server Runtime Args",
+                    "description": "You normally don't set this. Additional arguments to pass to the node executable when spawning the language server. This is useful when you use something like Yarn PnP and need its loader arguments `[\"--loader\", \".pnp.loader.mjs\"]`."
                 },
                 "svelte.language-server.port": {
                     "type": "number",

--- a/packages/svelte-vscode/src/extension.ts
+++ b/packages/svelte-vscode/src/extension.ts
@@ -111,22 +111,35 @@ export function activateSvelteLanguageServer(context: ExtensionContext) {
     // Add --experimental-modules flag for people using node 12 < version < 12.17
     // Remove this in mid 2022 and bump vs code minimum required version to 1.55
     const runExecArgv: string[] = ['--experimental-modules'];
-    let port = runtimeConfig.get<number>('port') ?? -1;
+
+    const runtimeArgs = runtimeConfig.get<string[]>('runtime-args');
+    if (runtimeArgs !== undefined) {
+        runExecArgv.push(...runtimeArgs);
+    }
+
+    const debugArgs = ['--nolazy'];
+
+    const port = runtimeConfig.get<number>('port') ?? -1;
     if (port < 0) {
-        port = 6009;
+        debugArgs.push('--inspect=6009');
     } else {
         console.log('setting port to', port);
         runExecArgv.push(`--inspect=${port}`);
     }
-    const debugOptions = { execArgv: ['--nolazy', '--experimental-modules', `--inspect=${port}`] };
+
+    debugArgs.push(...runExecArgv);
 
     const serverOptions: ServerOptions = {
         run: {
             module: serverModule,
             transport: TransportKind.ipc,
-            options: { execArgv: runExecArgv }
+            options: { execArgv: runExecArgv, cwd: rootPath }
         },
-        debug: { module: serverModule, transport: TransportKind.ipc, options: debugOptions }
+        debug: {
+            module: serverModule,
+            transport: TransportKind.ipc,
+            options: { execArgv: debugArgs, cwd: rootPath }
+        }
     };
 
     const serverRuntime = runtimeConfig.get<string>('runtime');

--- a/packages/svelte-vscode/src/extension.ts
+++ b/packages/svelte-vscode/src/extension.ts
@@ -133,12 +133,12 @@ export function activateSvelteLanguageServer(context: ExtensionContext) {
         run: {
             module: serverModule,
             transport: TransportKind.ipc,
-            options: { execArgv: runExecArgv, cwd: rootPath }
+            options: { execArgv: runExecArgv }
         },
         debug: {
             module: serverModule,
             transport: TransportKind.ipc,
-            options: { execArgv: debugArgs, cwd: rootPath }
+            options: { execArgv: debugArgs }
         }
     };
 

